module/state_synchronization/requester/execution_data_requester.go

```diff
@@ -0,0 +1,783 @@
+	// processed sequentially.
+	// However, this condition should not cause the node to crash as that would result in all nodes
+	// running this component to crash simultaneously. Pausing here keeps the component alive until
+	// shutdown.
+	<-ctx.Done()
+
+	// TODO: is there any cleanup to do after a halt?
+}
+
+func (e *executionDataRequesterImpl) fetchRequestProcessor(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	<-e.eds.Ready()
+	ready()
+
+	e.fetchRequestProcessingLoop(ctx)
+}
+
+func (e *executionDataRequesterImpl) notificationProcessor(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+	e.notificationProcessingLoop(ctx)
+}
+
+// finalizationProcessingLoop waits for finalized block notifications, then processes all available
+// blocks in the queue
+func (e *executionDataRequesterImpl) finalizationProcessingLoop(ctx irrecoverable.SignalerContext) {
+	for {
+		// prioritize shutdowns
+		if util.CheckClosed(ctx.Done()) || util.CheckClosed(e.stop) {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-e.stop:
+			return
+		case <-e.finalizedBlocksNotifier.Channel():
+			e.processAvailableBlocks(ctx)
+		}
+	}
+}
+
+func (e *executionDataRequesterImpl) processAvailableBlocks(ctx irrecoverable.SignalerContext) {
+	for {
+		item, ok := e.finalizedBlockQueue.Pop()
+
+		// return when queue is empty
+		if !ok {
+			return
+		}
+
+		blockID := item.(flow.Identifier)
+
+		block, err := e.blocks.ByID(blockID)
+
+		// block must be in the db, otherwise there's a problem with the state
+		if err != nil {
+			ctx.Throw(fmt.Errorf("failed to lookup finalized block %s in protocol state db: %w", blockID, err))
+		}
+
+		err = e.processBlockSeals(block)
+		if err != nil {
+			// all errors indicate a problem fetching data from the db
+			ctx.Throw(fmt.Errorf("failed to process finalized block %s: %w", blockID, err))
+		}
+	}
+}
+
+// Will return an error if any of the seals processed are for a block that does not exist in the db
+func (e *executionDataRequesterImpl) processBlockSeals(block *flow.Block) error {
+	logger := e.log.With().Str("finalized_block_id", block.ID().String()).Logger()
+	logger.Debug().Msgf("checking %d seals in block", len(block.Payload.Seals))
+
+	if len(block.Payload.Seals) == 0 {
+		return nil
+	}
+
+	// First, find all seals in the block and sort them by height (ascending)
+
+	fetchRequests := []fetchRequest{}
+
+	for _, seal := range block.Payload.Seals {
+		sealedBlock, err := e.blocks.ByID(seal.BlockID)
+
+		if err != nil {
+			return fmt.Errorf("failed to lookup sealed block in protocol state db: %w", err)
+		}
+
+		logger.Debug().
+			Str("sealed_block_id", seal.BlockID.String()).
+			Uint64("sealed_block_height", sealedBlock.Header.Height).
+			Msg("checking sealed block")
+
+		fetchRequests = append(fetchRequests, fetchRequest{
+			blockID: seal.BlockID,
+			height:  sealedBlock.Header.Height,
+		})
```

Comment:
Why don't we just directly fetch the Execution Data ID here and put that into the queue? That way the queue consumers (workers) can just fetch the given ID?

Reply:
since there are more fetch workers than finalization, I wanted to offload more processing to them. it's only a few ms though so it doesn't make a big difference